src/metabase/util/log.clj

```diff
@@ -48,17 +48,61 @@
          (lambdaisland.glogi/log logger# level# nil s#)
          a#))))
 
+(defmacro with-context
+  "Executes body with the given context map and message prefix in ThreadContext.
+   The context map's keys and values are added to the ThreadContext individually.
+   Preserves any existing context values and restores them after execution.
+
+   The keys in context-map should be keywords.
+
+   Example usage:
+   (with-context {:notification_id 1}
+     (log/infof \"Hello\"))
+
+   ThreadContext will contain: {\"notification_id\" \"1\"} and stack \"Notification 1\""
+  [context-map & body]
+  (macros/case
+    :clj `(let [ctx-map# ~context-map
+                ctx-keys# (keys ctx-map#)
+                ;; Store original values before modifying
+                original-values# (into {}
+                                       (keep (fn [k#]
+                                               (when-let [v# (ThreadContext/get (str (symbol k#)))]
+                                                 [(str (symbol k#)) v#])))
+                                       ctx-keys#)]
+            (try
+              (doseq [k# ctx-keys#]
+                (ThreadContext/put (str (symbol k#)) (str (get ctx-map# k#))))
+              ~@body
+              (finally
+                (doseq [k# ctx-keys#]
+                  (if-let [original# (find original-values# (str (symbol k#)))]
+                    (ThreadContext/put (str (symbol k#)) (val original#))
+                    (ThreadContext/remove (str (symbol k#))))))))
+    "cljs" ~@body))
+
+(defn parse-args
+  "Parses args for [[trace]], [[debug]], [[info]], [[warn]], [[error]], [[fatal]]
+  into a map with message, context and exception keys."
+  [& args]
+  (let [first-arg (first args)
+        has-exception? (instance? Exception first-arg)
+
+        rest-args (if has-exception? (rest args) args)
+        has-context? (map? (last rest-args))
+        message-args (cond-> rest-args has-context? drop-last)]
```

Comment:
absolutely negligible nitpick, but `(rest args)` could be in that `cond->` too 😁